QL: Simplify arithmetic operations in binary comps

bpintea opened the pull request:
This PR adds an optimizer rule to simplify the arithmetic operations
in binary comparison expressions, which in turn will allow for further
expression compounding by the optimiser.

Only the negation and plus, minus, multiplication and division are
currently considered and only when two of the operands are a literal.

For instance `((a + 1) / 2 - 3) * 4 >= 14` becomes `a >= 12`.

Closes #65394.

bpintea committed:
Simplify arithmetic operations in binary comps

This commit adds an optimizer rule to simplify the arithmetic operations
in binary comparison expressions, which in turn will allow for further
expression compounding by the optimiser.

Only the negation and plus, minus, multiplication and division are
currently considered and only when two of the operands are a literal.

For instance `((a + 1) / 2 - 3) * 4 >= 14` becomes `a >= 12`.

bpintea committed:
Undo unintentional change

Revert unintentional change of a source file.

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

astefan commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
+
+                if (opLeft.left() instanceof Literal || opLeft.right() instanceof Literal) { // 5 [op] a >= 4 || a [op] 5 >= 4
+                    // force double division
+                    if (opLeft.symbol().equals("*")) { // use symbol comp for SQL's Mul
```
Why do you identify multiplication with `wildcard` symbol and not with an `instanceof`?

bpintea replied:
The SQL ops aren't available as deps in the QL package.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/BinaryComparison.java:
```diff
@@ -49,4 +49,6 @@ protected Pipe makePipe() {
         return Comparisons.compare(left, right);
     }
+
+    public abstract BinaryComparison reverse();
```
Add javadoc regarding the purpose of the method.

bpintea replied:
javadoc added.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Mul.java:
```diff
@@ -52,4 +52,9 @@ protected Mul replaceChildren(Expression newLeft, Expression newRight) {
     }
+
+    @Override
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
```
return `Div` instead of `ArithmeticOperation`.

bpintea replied:
Fixed.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Sub.java:
```diff
@@ -27,4 +27,9 @@ public Sub(Source source, Expression left, Expression right) {
     }
+
+    @Override
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
```
return `Add` instead of `ArithmeticOperation`

bpintea replied:
Fixed.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Mod.java:
```diff
@@ -30,4 +30,10 @@ public Mod(Source source, Expression left, Expression right) {
     }
+
+    @Override
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
+        // TODO: Modular Multiplicative Inverse, if ever needed?
+        throw new UnsupportedOperationException("inverting modulo operation is not supported");
+    }
```
The exception type is wrong. And we rarely thrown an exception on methods that are forced to be implemented - that indicates either the code is incomplete (as here) or that the contract is too broad.

bpintea replied:
Fixed, adopted the same strategy as `ArithmeticOperation#swapLeftAndRight()`.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Div.java:
```diff
@@ -34,4 +34,9 @@ protected Div replaceChildren(Expression newLeft, Expression newRight) {
     }
+
+    @Override
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
```
return `Mul` instead of `ArithmeticOperation`

bpintea replied:
Fixed.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/ArithmeticOperation.java:
```diff
@@ -45,4 +45,6 @@ public DataType dataType() {
         return new BinaryArithmeticPipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), function());
     }
+
+    public abstract ArithmeticOperation inverse(Source source, Expression left, Expression right);
```
Javadoc on what this does.

bpintea replied:
Added a description.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Add.java:
```diff
@@ -30,4 +30,9 @@ protected Add replaceChildren(Expression left, Expression right) {
     }
+
+    @Override
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
```
return `Sub` instead of `ArithmeticOperation`

bpintea replied:
Fixed.

astefan reviewed (changes requested):
This PR is changing one behavior that might not be so obvious at first: a query like `SELECT salary FROM test_emp WHERE salary + 10 >= 'x'` returns no documents in `master`, but throws `Cannot compute [-] between [String] [Integer]` in this PR.

Apart from the fact that the error message is slightly incomplete - missing `and` between `[String]` and `[Integer]` - it's a confusing error message for the user. The minus sign is not present in the query, but present in the error message.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerRunTests.java:
```diff
@@ -48,4 +57,57 @@ public void testWhereClause() {
+
+    private void doTestBinaryComparisonReduction(String expression, Class<? extends BinaryComparison> binaryComparisonClass,
+                                                 Number bound) {
+        LogicalPlan plan = plan("SELECT some.string FROM test WHERE " + expression);
+
+        assertTrue(plan instanceof UnaryPlan);
+        UnaryPlan unaryPlan = (UnaryPlan) plan;
+        assertTrue(unaryPlan.child() instanceof Filter);
+        Filter filter = (Filter) unaryPlan.child();
+        assertEquals(binaryComparisonClass, filter.condition().getClass());
+        BinaryComparison bc = (BinaryComparison) filter.condition();
+
+        assertTrue(bc.left() instanceof FieldAttribute);
+        FieldAttribute attribute = (FieldAttribute) bc.left();
+        assertEquals("int", attribute.name());
+
+        assertTrue(bc.right() instanceof Literal);
```
The test only cares about the expression so you can just use `parser.createExpression` (see tests inside `ExpressionTests`).

bpintea replied:
I've employed `parser.createExpression()`, thanks.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerRunTests.java:
```diff
@@ -48,4 +57,57 @@ public void testWhereClause() {
-}
\ No newline at end of file
+
+    public void testReducedBinaryComparisonGreaterThenOrEqual() {
```
I'd like to see more tests that the behavior regarding commutative vs non-commutative expressions:
```
a + 2 > 3
2 + a > 3
a - 2 > 3
2 - a > 3
a * 2 > 3
2 * a > 3
a / 2 > 3
2 / a > 3
1 / (a - 2) > 3
1 / (2 - a) > 3
1 / (2 + (2 - a) > 3
```

bpintea replied:
Added all these tests.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
+
+        @Override
+        protected Expression rule(Expression e) {
+            if (e instanceof BinaryComparison) {
+                return reduce((BinaryComparison) e);
+            }
+            return e;
```
```suggestion
            if (e instanceof BinaryComparison) {
                e = reduce((BinaryComparison) e);
            }
            return e;
```

bpintea replied:
Refactored.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
+        }
+
+        private Expression reduce(BinaryComparison bc) {
+            // optimize only once the expression looks like: AnyNode [ArithmOp] AnyNode [BiComp] Literal
```
```suggestion
            // optimize only once the expression looks like `Expression [ArithmeticOp] Expression [BinaryComp] Literal`
```

bpintea replied:
Rephrased.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
+
+                if (opLeft.left() instanceof Literal || opLeft.right() instanceof Literal) { // 5 [op] a >= 4 || a [op] 5 >= 4
+                    // force double division
+                    if (opLeft.symbol().equals("*")) { // use symbol comp for SQL's Mul
```
Try to avoid using strings for type matching. Assuming there's no other either way move use a string reference to indicate its source, such as `DefaultBinaryArithmeticOperation.MUL.symbol()` instead of `*`.

Also please add some comments to explain what's going on - if the symbol is Mul, force double division by promoting the right literal to a double?

bpintea replied:
>  such as DefaultBinaryArithmeticOperation.MUL.symbol() instead of *.

Thanks, fixed.

> if the symbol is Mul, force double division by promoting the right literal to a double?

Yes, that was the logic. But this got now removed, as a potentially unsafe optimisation.

costin reviewed (changes requested):
I'd like to see more tests including non-numeric values considering the binary operators inside SQL support intervals and dates.
The cast to double is problematic - what does `4 / 2` return vs `5 / 2`? And why is the double casting having to be applied manually when we do not do that for `/` in the first place.
What happens in case of : 
```
date - INTERVAL 1 MONTH > '2010-01-01T01:01:01` ? 
```

palesz commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
+                    if (opLeft.symbol().equals("*")) { // use symbol comp for SQL's Mul
+                        bcRight = new Literal(bcRight.source(), ((Number) bcRight.value()).doubleValue(), DOUBLE);
+                    }
+                    return bc.replaceChildren(List.of(opLeft.left(), opLeft.inverse(bcRight.source(), bcRight, opLeft.right())));
```
Precision loss might cause some issues here. `a + b == c` is mathematically the same as `a == c - b`, except when you take into account the potential precision loss.

For example:
```
a = 1.0E-15
b = 1.0E15
c = 1.0E15

a + b == c
1.0E-15 + 1.0E15 == 1.0E15
1.0E15 == 1.0E15
true

a == c - b
1.0E-15 == 1.0E15 - 1.0E15
1.0E-15 == 0.0
false
```

So even though the query before and after the optimization might be mathematically equivalent, it still won't return the same result after the optimization is applied. This might have unintended consequences after a version upgrade. I think we should discuss this before moving ahead with the PR.

bpintea replied:
Thanks, good point. Addressed, along with the similar DIV/MUL concern.

palesz commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerRunTests.java:
```diff
@@ -48,4 +57,57 @@ public void testWhereClause() {
     }
-}
\ No newline at end of file
+
+    public void testReducedBinaryComparisonGreaterThenOrEqual() {
```
I would add some tests for testing with numbers where overflow or precision loss can occur.

bpintea replied:
I've added more tests.

palesz reviewed (changes requested):
I have one major concern that I would like to discuss before moving ahead with the PR.

bpintea committed:
Refactor the optimisation. Adress review comments

The optimisation is now refactor splitting the logic by add-sub and
mul-div operations.
Further restictions are put in place, to prevent the
optimiser from chaning the outcome of the filtering.
More tests have been added.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/optimize_arithm_in_bc

bpintea committed:
Style fix

Removed now unused import.

bpintea committed:
Test fix

- update test condition to remove (non-optimisable) integer division.

bpintea commented:
> This PR is changing one behavior that might not be so obvious at first: a query like `SELECT salary FROM test_emp WHERE salary + 10 >= 'x'` returns no documents in `master`, but throws `Cannot compute [-] between [String] [Integer]` in this PR.

Thanks, I was wrongly expecting that this check had already been in place (but #49636 is still open).

> Apart from the fact that the error message is slightly incomplete - missing `and` between `[String]` and `[Integer]` - it's a confusing error message for the user. The minus sign is not present in the query, but present in the error message.

This hasn't been reported before, so I went ahead and fixed it part of this PR.

palesz commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+            return bc;
+        }
+
+        private static Expression reduceNegation(BinaryComparison bc) {
+            Literal bcLiteral = (Literal) bc.right();
+            Expression neg = safeMaybeFold(new Neg(bcLiteral.source(), bcLiteral));
+            return neg == null ? bc : bc.reverse().replaceChildren(List.of(bc.left().children().get(0), neg));
+        }
```
Some renaming and inlining would make this easier to read and reason about. The `safeMaybeFold` is just a really simple `try-catch` that you kind of mimic with the `null` check anyways.

```suggestion
        private static Expression reduceNegation(BinaryComparison bc) {
            Literal oldRight = (Literal) bc.right();
            Neg newRight = new Neg(oldRight.source(), oldRight);
            Expression newLeft = ((Neg) bc.left()).field();
            try {
                newRight = new Literal(newRight.source(), newRight.fold(), newRight.dataType());
            } catch (Exception e) {
                // cannot optimize the negation, either ArithmeticException or DateTimeException
                // but safe to just turn off the optmisation should any exception arise
                return bc;
            }
            return bc.reverse().replaceChildren(List.of(newLeft, newRight));
        }
```

bpintea replied:
`safeMaybeFold` performs an additional check too and is called from other places too, so I'd personally prefer the existing more compact and code reusing way.

palesz commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+            }
+
+            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                ArithmeticOperation operation = (ArithmeticOperation) comparison.left();
+                // Use symbol comp: SQL operations aren't available in this package (as dependencies)
+                String opSymbol = operation.symbol();
+                // Modulo can't be simplified.
+                if (MOD.symbol().equals(opSymbol)) {
+                    return comparison;
+                }
+                SimplifyOperation simplification = (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol))
+                    ? new SimplifyMulDiv(comparison)
+                    : new SimplifyAddSub(comparison);
```
Remove the explicit check on the not handled cases (MUL), just to stay on the safe side in case we add any sort of magic Arithmetic operators (so any possible new operation won't be treated as Add/Sub by default).

```suggestion
                SimplifyOperation simplification = null;
                if (isMulOrDiv(comparison.left()) {
                    simplification = new SimplifyMulDiv(comparison);
                } else if (isAddOrSub(comparison.left()) {
                    simplification = new SimpliAddSub(comparison);
                }
                if (simplification == null) {
                    return comparison;
                }
```

bpintea replied:
I was wrongly thinking mostly at SQL (and less at QL) and assuming we won't add operators (without at least breaking these tests), but sure, it absolutely makes sense to expand the check.

palesz commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Mod.java:
```diff
@@ -12,7 +12,7 @@
+ *
  * Note this operator is also registered as a function (needed for ODBC/SQL) purposes.
  */
 public class Mod extends ArithmeticOperation {
```
I think the `inverse` should throw an exception in this class.

bpintea replied:
See above [comment](https://github.com/elastic/elasticsearch/pull/66022#discussion_r553452382).

palesz reviewed (commented):
Halfway through, few comments and suggestions so far.

costin reviewed (changes requested):
Looks to me like this PR is made of two distinct parts.
1. the 'bubbling' of negations (which I would rename to SimplifyDoubleNegations).
2. The actual simplification of arithmetic in comparisons.

2 depends on 1 however 1 has value by itself and it would be easier to reason about stand-alone.

bpintea committed:
Address review comments

- add isMulOrDiv and isAddOrSub methods;
- add a new factory if'ace for the arithmetic operations invertible in
  relation to a binary comparison;
- simplify NEG - NEG tree structures;
- more tests.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/ArithmeticOperation.java:
```diff
@@ -45,4 +45,11 @@ public DataType dataType() {
     protected Pipe makePipe() {
         return new BinaryArithmeticPipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), function());
     }
+
+    /**
+     * Returns the opposite of this operation.
+     */
+    public ArithmeticOperation inverse(Source source, Expression left, Expression right) {
+        return this;
+    }
```
This had been implemented as `abstract` initially (and have `Mod` throw an exception, as you were also proposing below), but got refactored not to throw an exception (see https://github.com/elastic/elasticsearch/pull/66022#discussion_r538432770).
And after https://github.com/elastic/elasticsearch/pull/66022#discussion_r552012954 it got a second refactoring, as that comment addresses the approach.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                if (child instanceof ArithmeticOperation) {
+                    ArithmeticOperation operation = (ArithmeticOperation) child;
+                    String opSymbol = operation.symbol();
+                    if (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol)) {
```
Extracted, thanks.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+            } else if (e instanceof ArithmeticOperation) { // pulls up MUL/DIV - NEG - x ==> NEG - MUL/DIV - x
+                ArithmeticOperation operation = (ArithmeticOperation) e;
+                String opSymbol = operation.symbol();
+                if (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol)) {
```
Fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                    String opSymbol = operation.symbol();
+                    if (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol)) {
+                        if (operation.left() instanceof Neg) {
+                            return operation.replaceChildren(List.of(operation.left().children().get(0), operation.right()));
```
Good point, thanks.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                        }
+                        if (operation.right() instanceof Neg) {
+                            return operation.replaceChildren(List.of(operation.left(), operation.right().children().get(0)));
+                        }
```
Fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                if (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol)) {
+                    if (operation.left() instanceof Neg) {
+                        Neg neg = (Neg) operation.left();
+                        return new Neg(operation.source(), operation.replaceChildren(List.of(neg.children().get(0), operation.right())));
```
Fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                    }
+                    if (operation.right() instanceof Neg) {
+                        Neg neg = (Neg) operation.right();
+                        return new Neg(operation.source(), operation.replaceChildren(List.of(operation.left(), neg.children().get(0))));
```
Fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
```
Double negation is taken care of by `SimplifyComparisonsArithmetics#reduceNegation()` -- the point of this rule is really to surface the negations, not necessarily simplify them (which however happens as a by product). 
However, at Costin's [suggestion](https://github.com/elastic/elasticsearch/pull/66022#pullrequestreview-561895753) I'll extract this rule in separate PR, which clearly warrants having this simplification part of the rule.

bpintea commented on x-pack/plugin/eql/src/test/resources/queryfolder_tests.txt:
```diff
@@ -454,35 +454,35 @@ process where wildcard(process_path, "*\\red_ttp\\wininit.*", "*\\abc\\*", "*def
 
 addOperator
-process where serial_event_id + 2 == 41
+process where serial_event_id + 2 == -2147483647
```
They were changed so that the arithmetics optimisation won't kick in and so that the scripted translation can still be verified by the test (the new values would cause an overflow if optimised).

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/ArithmeticOperation.java:
```diff
@@ -45,4 +45,11 @@ public DataType dataType() {
         return new BinaryArithmeticPipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), function());
     }
+
+    /**
```
Indeed, this makes sense, thanks.
I've removed the `inverse()` method from the `ArithmeticOperation` class and added a clearer interface implemented only by those operations that are actually invertible in relation to the binary comparison.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Add.java:
```diff
@@ -30,4 +30,9 @@ protected Add replaceChildren(Expression left, Expression right) {
     public Add swapLeftAndRight() {
         return new Add(source(), right(), left());
     }
+
```
Refactored.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                try {
+                    expression = new Literal(expression.source(), expression.fold(), expression.dataType());
+                } catch (Exception e) {
+                    // could only catch (ArithmeticException | DateTimeException e), but safer to just turn off the optimisation
```
Fixed.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec:
```diff
@@ -87,3 +87,149 @@ orderByModulo
 SELECT emp_no FROM test_emp ORDER BY emp_no % 10000 LIMIT 10;
 orderByMul
 SELECT emp_no FROM test_emp ORDER BY emp_no * 2 LIMIT 10;
+
```
All these tests are done in `x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerRunTests.java`, but I've added a few more here too (min/max long values and double negations were already present).

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -2074,15 +2074,15 @@ public void testNoCountDoesNotTrackHits() {
         assertThat(eqe.queryContainer().toString().replaceAll("\\s+", ""), containsString(
                 "\"script\":{\"script\":{\"source\":\"InternalQlScriptUtils.nullSafeFilter("
-                        + "InternalQlScriptUtils.gt(InternalSqlScriptUtils.add(InternalQlScriptUtils.docValue(doc,params.v0),"
+                        + "InternalQlScriptUtils.gt(InternalSqlScriptUtils.sub(InternalQlScriptUtils.docValue(doc,params.v0),"
```
I've changed the test from an addition to a subtraction, as well as used a max value, so that the optimisation is prevented. Consequently the translation had to be adjusted.

costin replied:
Could you then add a test where the optimization is applied? I'm confused since in this instance, I expect the PR to add tests not modify existing ones.

bpintea replied:
:-) Sure, tests have been added.
This test however still needs to be changed (as is now), as it meant to test `testZonedDateTimeInScripts`, not optimisations. What it did was the integer similar of `x + 2  > 3` which the new optimisation rule would translate to `x > 1`; so it would simplify the test, reducing its testing value. My change only made sure the literals non-field values are such that the optimisation won't take place, so that the test retains its original intention (sort of `x - MAX > 3`, just like [here](https://github.com/elastic/elasticsearch/pull/66022#discussion_r553453278)).

bpintea commented:
> 1. the 'bubbling' of negations (which I would rename to SimplifyDoubleNegations).

I've converted the PR back to draft to extract the rule on negations. However, I assume the rule will mostly extract the negation up (i.e. bubbling them up: DIV/MUL - NEG - x => NEG - DIV/MUL - x), rather then reduce them (i.e. NEG - NEG - x => x), so not sure if "Simplify" would be correct.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
Remove negation optmisations

These are unsafe, see
https://github.com/elastic/elasticsearch/pull/67278

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
Reintroduce negation reduction

* `NEG - x < Literal` becomes `x > -Literal` if folding can be done.
* add more tests

bpintea committed:
Revert WS-only changes

Revert changes in files where only WS was updated.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
CSV spec test

- add test on negation handling with integer limit value.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+
+            @Override
+            Expression doSimplify() {
+                if (SUB.symbol().equals(operation.symbol()) && opRight instanceof Literal == false) {
```
Check the enums directly, don't use their names.

costin replied:
Also is this condition correct? the else (parent doSimplify) is picked if it is a sub but the opRight is a literal

bpintea replied:
> Check the enums directly, don't use their names.

See above.

> Also is this condition correct? the else (parent doSimplify) is picked if it is a sub but the opRight is a literal

It is correct: in the else-case the literal has simply been "normalised" to the right and the simplification can be attempted.
The check you're referring to (in the child `SimplifyAddSub`) is there to avoid transforming the expression that ultimately still can't be simplified; ex.: for `1 - x > -MAX`, w/o the check, the next transformation would be `x - MAX < 1` and then it would stop there, since `MAX + 1` would overflow. That's what the comment on the first line below the last in the snippet above indicates.
But if the literal is on the right already (`x - 1 ...`), the optimisation can be attempted in the parent `doSimplify` (as per previous naming).

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+                return typesCompatible.apply(bcLiteral.dataType(), opLiteral.dataType()) == false;
+            }
+
+            Expression doSimplify() {
```
Looks like this method is never called in this form, subclasses always override it and enhance it. If that's the case, simply declare the method as is and declare a separate one as abstract which subclasses are forced to override and in where they can call this method.

bpintea replied:
The method was called in that form, but anyways, code got refactored as per the feedback.

costin reviewed (changes requested):
This looks better however `SimplifyOperation` needs improving.
It's a mixture of parent plus children classes with static methods and templating making it hard to follow.

Extract the `simplify` static method from the SimplifyOperation to make it stand-alone so the  class only deals with the hierarchy. The naming could be improved:
`simplify == null || simplify.cannotSimplify() ? .. : simplify.doSimplify()` - fairly repetitive without conveying much.

The idea is there but the wording can be better, e.g.:
`operation/simplification == null || simplification.checkExpression/canBeApplied() ? simplification.apply()`.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+    public static final class SimplifyComparisonsArithmetics extends OptimizerExpressionRule {
+        BiFunction<DataType, DataType, Boolean> typesCompatible;
+
+        public SimplifyComparisonsArithmetics(BiFunction<DataType, DataType, Boolean> typesCompatible) {
```
`typesCompatible` -> `typeCheck`

bpintea replied:
Before I change it, just to 2x check: the passed arguments are `DataTypes::areCompatible` and `SqlDataTypes::areCompatible`.
The code using the var also looks like `return typesCompatible.apply([type1], [type2]) == false;`.
Not sure if the change would bring more clarity ("what's checked?"), so asking.

astefan reviewed (commented):
Left some comments

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
Address feedback

- rename methods for more clarify;
- regroup simplification checks;
- eliminate calling super class' methods;
- use == for static string comps;
- use Math.signum instead of inlining the calculation.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/optimize_arithm_in_bc

bpintea committed:
Address review comments

Update QA test.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Add.java:
```diff
@@ -12,7 +12,7 @@
  * Addition function ({@code a + b}).
  */
-public class Add extends DateTimeArithmeticOperation {
+public class Add extends DateTimeArithmeticOperation implements BinaryComparisonInvertible {
```
I've changed it to Inversible..

Since the interface is meant to convey the action intent, I find `-ble` needed. But FTR, `invertible` might still be more appropriate since "inversible" isn't really a word and a word already exists: that's why a matrix^-1 is "inverted" and so are also indexes/indices ("_inverted_ index" not "inversed index").

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1284,7 +1246,12 @@ Expression doSimplify() {
-                Expression bcRightExpression = safeMaybeFold(operation.inverse(bcl.source(), bcl, opRight));
+                if (operation instanceof BinaryComparisonInvertible == false) { // should be an assertion?
+                    throw new QlIllegalArgumentException("Unexpected [" + operation.symbol() + "] operation object");
+                }
```
I've removed the check.

Offtopic:
> no need to double check the code flow

Extensive testing should definitely help. I believe assertions exist still because code has a (life) cycle and effects of later changes can't always be exhaustively checked/predicted or coded in (see the many `instanceof`s or [comment](https://github.com/elastic/elasticsearch/pull/66022#discussion_r557686188)).

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1148,46 +1152,6 @@ protected Expression rule(Expression e) {
-    }
-
     // Simplifies arithmetic expressions with fixed point fields in BinaryComparisons.
     public static final class SimplifyComparisonsArithmetics extends OptimizerExpressionRule {
```
Code refactored and comments added.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(List.of(((Neg) bc.left()).field(), literalNeg));
+        }
+
+        static Expression safeMaybeFold(Expression expression) {
```
Suggestion applied.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+        private static Expression reduceNegation(BinaryComparison bc) {
+            Literal bcLiteral = (Literal) bc.right();
+            Expression literalNeg = safeMaybeFold(new Neg(bcLiteral.source(), bcLiteral));
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(List.of(((Neg) bc.left()).field(), literalNeg));
```
Right, thanks.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            return bc;
+        }
+
+        private static Expression reduceNegation(BinaryComparison bc) {
```
Suggestion applied.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            return expression;
+        }
+
+        private static class SimplifyOperation {
```
It's been made abstract.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+
+            SimplifyOperation(BinaryComparison comparison) {
+                this.comparison = comparison;
+                operation = (ArithmeticOperation) comparison.left();
+                bcLiteral = (Literal) comparison.right();
```
Yes, all are checked upstream.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+                return bcRightExpression == null ? comparison : comparison.replaceChildren(List.of(opLeft, bcRightExpression));
+            }
+
+            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
```
The intention was to have the (previously called) SimplifyOperation be self-contained: deal with simplification operations entirely and not leak details into the outer class, that should not care if it's a AddSub or DivMul operation.
Anyways, I've extracted it out.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+                    simplify = new SimplifyMulDiv(comparison);
+                } else if (ADD.symbol().equals(opSymbol) || SUB.symbol().equals(opSymbol)) {
+                    simplify = new SimplifyAddSub(comparison);
+                }
```
The SQL operations that this code also acts on aren't available (in the current package) -- there's a comment on the line right before the first in the snippet above. So I don't know how else it can be done.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+                int sign = 1;
+                if (obj instanceof Number) {
+                    double d = ((Number) obj).doubleValue();
+                    sign = d == 0 ? 0 : (d < 0 ? -1 : 1);
```
Thanks, fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            @Override
+            boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                // no ADD/SUB with floating fields
+                return operation.dataType().isRational() || super.cannotSimplify(typesCompatible);
```
Separate methods declared.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(List.of(((Neg) bc.left()).field(), literalNeg));
+        }
+
+        static Expression safeMaybeFold(Expression expression) {
```
> I think the return null inside the catch block bothers me.

The `null` isn't returned from the catch block, it's just assigned to the var to return.
But this still being a detail, would you mind expanding what's troubling / how it could be improved?

> I would have kept the original (no separate method used)

The code is extracted as a method since it's reused. It should be acceptably concise too: "if the expression is foldable, attempt to do so. Return either folded outcome, original (if not foldable) or `null` (if foldable, but unsafely; i.e. exception wud be thrown)".
An alternative could be for the caller to save the input expression (as in your example), the function to return the original input, if folding is unsafe, and for caller to check returned value against the original input. But this would be more verbose (as in your example) and this pattern to be applied multiple times.

For now I've taken Costin's [suggestion](https://github.com/elastic/elasticsearch/pull/66022#discussion_r557683449), but still open for change if needed.

astefan replied:
Can be left as is I guess. My initial suggestion seems to be just a personal preference. If there will be other changes to this PR, I'd add a comment similar to the one I made in the `catch (ArithmeticException | DateTimeException e) {` to explain the catch block "ignoring" the math failure.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+            if (bc.right() instanceof Literal) {
+                if (bc.left() instanceof ArithmeticOperation) {
+                    return SimplifyOperation.simplify(bc, typesCompatible);
+                } else if (bc.left() instanceof Neg) {
```
I've changed it.

(Particularly here, the if[-else] is practically a switch, on a type that's not switchable in Java, so the change makes sense.
_Generally_ I'm in favour of the if-else format when:
* a `return` is involved, whose later replacement by a non-jumping statement, could make the next `if` branch eligible; i.e. less error-prone.
* is more compact.)

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+
+            SimplifyMulDiv(BinaryComparison comparison) {
+                super(comparison);
+                isDiv = DIV.symbol().equals(operation.symbol());
```
Indeed, `==` should be safe for as long as we don't assign dynamic values to operations' symbols. I've changed them.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.csv-spec:
```diff
@@ -23,7 +23,51 @@ nullArithmetics
+;
+
+optimizedIntervalFilterPlus
+SELECT emp_no x FROM test_emp WHERE hire_date + INTERVAL 20 YEAR > CAST('2010-01-01T00:00:00' AS TIMESTAMP) LIMIT 10;
```
Makes sense, added.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.csv-spec:
```diff
@@ -23,7 +23,51 @@ nullArithmetics
+;
+
+optimizedIntervalFilterMinus
+SELECT emp_no x FROM test_emp WHERE hire_date - INTERVAL 10 YEAR > CAST('1980-01-01T00:00:00' AS TIMESTAMP) LIMIT 10;
```
Added.

astefan reviewed (approved):
LGTM

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1170,21 +1171,44 @@ private Expression simplify(BinaryComparison bc) {
+            return (simplification == null || simplification.isUnsafe(typesCompatible)) ? comparison : simplification.apply();
+        }
+
+        private static boolean isMulOrDiv(String opSymbol) {
```
If this method necessary? I only see it used in one place and the check can check can be directly just life for ADD and SUB.

bpintea replied:
It's being used twice. I'm on the fence tbh - it makes things easier to read in `sign()`, but looks unbalanced in the if-else explicit ADD/SUB check - but would leave it as is now.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1237,109 +1261,97 @@ boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible)
-            static boolean isMulOrDiv(String opSymbol) {
-                return MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol);
+            //  - post adjustments
+            Expression opSpecificPostApply(BinaryComparison binaryComparison) {
```
postApply or postProcess. opSpecific is somewhat meaningless.

bpintea replied:
I went for `postProcess`.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1221,7 +1245,7 @@ static Expression safeMaybeFold(Expression expression) {
 
             // can it be quickly fast-tracked that the operation can't be reduced?
-            boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+            boolean isUnsafe(BiFunction<DataType, DataType, Boolean> typesCompatible) {
```
If you don't this to be subclass because of the dedicated method, consider making it final.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -1237,109 +1261,97 @@ boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible)
+                    return true;
+                }
+
+                return opSpecificIsUnsafe();
```
how about `check`, `doCheck` without `opSpecific`?

bpintea replied:
Updated.

costin reviewed (approved):
LGTM. Left some stylist comments, no need for another review from me.

bpintea committed:
Address review comments

- rename methods;
- add comments.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/optimize_arithm_in_bc

bpintea committed:
Adjust merging outcome for one test

Both upstream and current branch updated line in the same way
(incrementing a count), so merging left it unchanged. However the net
result needs to reflect both modifications (i.e. inc by two).

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/optimize_arithm_in_bc
